File: Trees/fenwick_tree.cpp
#include <bits/stdc++.h>

using namespace std;

// Operations: O(log n)
// Space: O(n)
class FenwickTree {
private:
    vector<int> tree;

    int LSB(int i) {
        return i & -i;
    }

public:
    FenwickTree(int size) {
        tree.resize(size+1, 0);
    }

    // Construct Fenwick Tree
    FenwickTree(const vector<int>& nums) {
        int n = (int)nums.size();
        tree.resize(n+1, 0);

        for(int i = 0; i < n; i++) {
            update(i, nums[i]);
        }
    }

    // O(n) Fenwick initialization
    void init(int n, const vector<int>& nums) {
        for (int i = 1; i <= n; i++) {
            tree[i] += nums[i];
            if (i + LSB(i) <= n) {
                tree[i + LSB(i)] += tree[i];
            }
        }
    }

    void update(int index, int value) {
        index++;
        while(index < (int)tree.size()) {
            tree[index] += value;
            index += LSB(index);
        }
    }

    // Get prefix sum [0, index]
    int sum_query(int index) {
        index++;
        int sum = 0;
        while(index > 0) {
            sum += tree[index];
            index -= LSB(index);
        }
        return sum;
    }

    // Get range sum [l, r] (same as prefix sum formula)
    int range_sum_query(int l, int r) {
        return sum_query(r) - sum_query(l-1);
    }
};


// testing example
int main(void) {
    vector<int> nums = {1, 3, 5, 7, 9, 11};
    FenwickTree ft(nums);

    cout << "Prefix sum [0, 3]: " << ft.sum_query(3) << "\n";        // 1+3+5+7 = 16
    cout << "Range sum [2, 4]: " << ft.range_sum_query(2, 4) << "\n"; // 5+7+9 = 21

    ft.update(3, 2);  // add 2 to element at index 3 (7+2 = 9)
    cout << "After update, prefix sum [0, 3]: " << ft.sum_query(3) << "\n";  // 1+3+5+9 = 18



    return 0;
}

File: datastructures/gol/fenwick/fenwick_tree.go
package fenwick

import "math/bits"

// Operations: O(log n)
// Space: O(n)
type FenwickTree struct {
	tree []int
}

func (f *FenwickTree) lsb(i int) int {
	return i & -i
}

func NewFenwickTree(size int) *FenwickTree {
	return &FenwickTree{tree: make([]int, size+1)}
}

// Construct Fenwick Tree.
func NewFenwickTreeFromNums(nums []int) *FenwickTree {
	ft := NewFenwickTree(len(nums))
	for i, x := range nums {
		ft.Update(i, x)
	}
	return ft
}

// O(n) Fenwick initialization.
func (f *FenwickTree) Init(n int, nums []int) {
	f.tree = make([]int, n+1)
	for i := 1; i <= n; i++ {
		f.tree[i] += nums[i]
		if i+f.lsb(i) <= n {
			f.tree[i+f.lsb(i)] += f.tree[i]
		}
	}
}

func (f *FenwickTree) Update(index, value int) {
	index++
	for index < len(f.tree) {
		f.tree[index] += value
		index += f.lsb(index)
	}
}

// Get prefix sum [0, index].
func (f *FenwickTree) SumQuery(index int) int {
	index++
	sum := 0
	for index > 0 {
		sum += f.tree[index]
		index -= f.lsb(index)
	}
	return sum
}

// Get range sum [l, r] (same as prefix sum formula).
func (f *FenwickTree) RangeSumQuery(l, r int) int {
	return f.SumQuery(r) - f.SumQuery(l-1)
}

// 0-based index of first prefix sum >= x.
// prefix sums have to be monotone for this.
func (f *FenwickTree) LowerBound(x int) int {
	pos := 0
	sum := 0
	for i := bits.Len(uint(len(f.tree))) - 1; i >= 0; i-- {
		nxt := pos + (1 << i)
		if nxt < len(f.tree) && sum+f.tree[nxt] < x {
			sum += f.tree[nxt]
			pos = nxt
		}
	}
	return pos
}
